packages/replay/src/upload.ts

```diff
@@ -8,278 +8,259 @@ import { Options, OriginalSourceEntry, RecordingMetadata, SourceMapEntry } from
+    const error = await waiter.promise;
+    return error;
+  }
 
-  gClient
-    .sendCommand("Session.ensureProcessed", { level: "basic" }, null, sessionId)
-    .then(() => waiter.resolve(null));
+  async connectionReportCrash(data: any) {
+    if (!this.client) throw new Error("Protocol client is not initialized");
 
-  const error = await waiter.promise;
-  return error;
-}
+    await this.client.sendCommand("Internal.reportCrash", { data });
+  }
 
-async function connectionReportCrash(data: any) {
-  if (!gClient) throw new Error("Protocol client is not initialized");
+  async connectionUploadRecording(recordingId: string, path: string) {
+    if (!this.client) throw new Error("Protocol client is not initialized");
 
-  await gClient.sendCommand("Internal.reportCrash", { data });
-}
+    const file = fs.createReadStream(path);
 
-// Granularity for splitting up a recording into chunks for uploading.
-const ChunkGranularity = 1024 * 1024;
+    file.on("error", e => {
+      console.error(`Failed to read replay ${recordingId} from disk`);
+      console.error(e);
 
-async function connectionUploadRecording(recordingId: string, path: string) {
-  if (!gClient) throw new Error("Protocol client is not initialized");
+      throw e;
+    });
 
-  const endWaiter = defer();
-  const promises = [];
-  const file = fs.createReadStream(path);
+    let buffer: Buffer | undefined;
+    let offset = 0;
 
-  let buffer: Buffer | undefined;
-  let offset = 0;
+    debug("%s: Replay file size: %d bytes", recordingId, fs.statSync(path).size);
 
-  debug("Replay file size: %d bytes", fs.statSync(path).size);
+    const send = async (b: Buffer) => {
+      const length = b.length;
+      debug("%s: Sending %d bytes at offset %d", recordingId, length, offset);
 
-  const send = (b: Buffer) => {
-    const length = b.length;
-    debug("Sending %d bytes at offset %d", length, offset);
+      await new Promise<void>((resolve, reject) =>
+        this.client?.sendCommand(
+          "Internal.addRecordingData",
+          { recordingId, offset, length },
+          b,
+          undefined,
+          err => (err ? reject(err) : resolve())
+        )
+      );
 
-    promises.push(
-      gClient?.sendCommand("Internal.addRecordingData", { recordingId, offset, length }, b)
-    );
+      offset += length;
+    };
 
-    offset += length;
-  };
+    for await (const chunk of file) {
+      const cb = chunk instanceof Buffer ? chunk : Buffer.from(chunk);
+      debug("%s: Read %d bytes from file", recordingId, cb.length);
 
-  file.on("data", chunk => {
-    const cb = chunk instanceof Buffer ? chunk : Buffer.from(chunk);
-    buffer = buffer ? Buffer.concat([buffer, cb]) : cb;
+      buffer = buffer ? Buffer.concat([buffer, cb]) : cb;
 
-    if (buffer.length >= ChunkGranularity) {
-      const data = buffer.subarray(0, ChunkGranularity);
-      buffer = buffer.subarray(ChunkGranularity);
-      send(data);
+      if (buffer.length >= ChunkGranularity) {
```

Comment:
I think this should maybe be a `while` instead of `if` since `buffer` could have enough data to be multiple chunks

Reply:
I don't think that'll be the case. The default `highwaterMark` is 16384 and `ChunkGranularity` is `1048576` so we should be always doing multiple reads for a chunk (unless there's somehow < 1 chunk which will be caught by the subsequent `send()`).

Reply:
ahh yeah fair enough!